File: internal/finder/list.go
package finder

import (
	"sort"
	"strings"

	tea "charm.land/bubbletea/v2"
	"charm.land/lipgloss/v2"
	"github.com/sahilm/fuzzy"

	"github.com/jedipunkz/gm/internal/repo"
)

// item is one row: a repository in the main list, a worktree in the Ctrl-W
// list. label is what is drawn and matched, path is what Enter yields.
type item struct {
	label string
	path  string
	score float64    // frecency; zero for worktrees
	seen  repo.Visit // the visit log; empty for worktrees
	// branch is the branch a row of the branch list stands for. Such a row
	// has a path only when the branch is already checked out somewhere.
	branch repo.Branch
	// pr is the pull request a row of the pull request list stands for, with
	// a path the same way.
	pr repo.PullRequest
}

type source []item

func (s source) String(i int) string { return s[i].label }
func (s source) Len() int            { return len(s) }

// dirtyMsg carries the result of a scan back to the UI thread.
type dirtyMsg map[string]bool

func (m *model) filter() {
	m.matched = map[int][]int{}
	m.note = ""
	// A command is not a query: the list stays as the query before it left it.
	q, _, _ := splitInput(strings.TrimSpace(m.input.Value()))
	q = strings.TrimSpace(q)
	// The selection follows the item, not the row number, whenever the query
	// is not making a new ranking statement — a command being typed, or the
	// query being cleared. Clearing it has to hold the selection still, or
	// there is no way to find a repository and then act on it.
	hold := m.view != nil && (q == m.query || q == "")
	held := -1
	if hold && m.cursor >= 0 && m.cursor < len(m.view) {
		held = m.view[m.cursor]
	}
	m.query = q
	if q == "" {
		m.view = make([]int, len(m.all))
		for i := range m.all {
			m.view[i] = i
		}
	} else {
		lower := strings.ToLower(q)
		matches := fuzzy.FindFrom(q, source(m.all))
		score := make(map[int]float64, len(matches))
		idx := make([]int, 0, len(matches))
		for _, mt := range matches {
			rel := m.all[mt.Index].label
			// fuzzy finds candidates; where it landed the characters is its
			// own guess, and a literal hit beats that guess every time.
			pos := substringMatch(rel, lower)
			if pos == nil {
				pos = mt.MatchedIndexes
			}
			score[mt.Index] = matchScore(rel, pos)
			m.matched[mt.Index] = runeIndexes(rel, pos)
			idx = append(idx, mt.Index)
		}
		// Ascending, so the strongest match lands at the bottom; frecency
		// breaks ties between equally good matches.
		sort.SliceStable(idx, func(a, b int) bool {
			if score[idx[a]] != score[idx[b]] {
				return score[idx[a]] < score[idx[b]]
			}
			return m.all[idx[a]].score < m.all[idx[b]].score
		})
		m.view = idx
	}
	m.view = m.keepDirty(m.view)

	if held >= 0 {
		for i, idx := range m.view {
			if idx == held {
				m.cursor = i
				return
			}
		}
	}
	m.cursor = len(m.view) - 1
}

// drop takes a removed repository out of the list, so the screen matches the
// disk without walking the tree again.
func (m *model) drop(path string) {
	kept := make([]item, 0, len(m.all))
	for _, it := range m.all {
		if it.path != path {
			kept = append(kept, it)
		}
	}
	m.all = kept
	delete(m.status, path)
	delete(m.probing, path)
	m.view = nil
	m.filter()
}

// add puts a new repository at the bottom of the list and selects it: it is
// the one thing the user is certain to want next.
func (m *model) add(rel, path string) {
	m.all = append(m.all, item{label: rel, path: path})
	m.view = nil
	m.input.SetValue("")
	m.filter()
	for i, idx := range m.view {
		if m.all[idx].path == path {
			m.cursor = i
		}
	}
}

// keepDirty drops the rows that have no uncommitted work, when the filter is
// on. It applies to repositories only: the worktree list is a different
// question, and a scan that has not finished yet hides nothing.
func (m model) keepDirty(view []int) []int {
	if !m.dirtyOnly || m.mode != modeRepos || m.dirty == nil {
		return view
	}
	kept := make([]int, 0, len(view))
	for _, i := range view {
		if m.dirty[m.all[i].path] {
			kept = append(kept, i)
		}
	}
	return kept
}

// scanDirty asks git about every repository at once, off the UI thread.
func (m model) scanDirty() tea.Cmd {
	paths := make([]string, 0, len(m.all))
	for _, it := range m.all {
		paths = append(paths, it.path)
	}
	scan := m.dirtyOf
	return func() tea.Msg { return dirtyMsg(scan(paths)) }
}

// runeIndexes converts fuzzy's byte offsets into rune positions, which is what
// rendering counts in.
func runeIndexes(s string, byteIdx []int) []int {
	if len(byteIdx) == 0 {
		return nil
	}
	want := make(map[int]bool, len(byteIdx))
	for _, b := range byteIdx {
		want[b] = true
	}
	out := make([]int, 0, len(byteIdx))
	ri := 0
	for bi := range s {
		if want[bi] {
			out = append(out, ri)
		}
		ri++
	}
	return out
}

// renderRow draws one repository, highlighting the characters the query
// matched and, when selected, the whole line.
func (m model) renderRow(i int, selected bool, width int) string {
	it := m.all[m.view[i]]
	base, hit, marker := m.st.Row, m.st.Hit, "  "
	if selected {
		base, hit, marker = m.st.RowSel, m.st.HitSel, m.st.Marker.Render("▸ ")
	}

	label := highlight(it.label, m.matched[m.view[i]], width-2, base, hit)
	gap := width - 2 - lipgloss.Width(label)
	if gap > 0 {
		label += base.Render(strings.Repeat(" ", gap))
	}
	return marker + label
}

// highlight renders s truncated to width, with the runes at hits in their own
// style. The tail is kept: the repository name matters more than the host.
func highlight(s string, hits []int, width int, base, hit lipgloss.Style) string {
	runes := []rune(s)
	off, prefix := 0, ""
	if width <= 1 {
		return ""
	}
	if len(runes) > width {
		off = len(runes) - width + 1
		prefix = "…"
		runes = runes[off:]
	}
	isHit := make(map[int]bool, len(hits))
	for _, h := range hits {
		isHit[h-off] = true
	}

	var b strings.Builder
	if prefix != "" {
		b.WriteString(base.Render(prefix))
	}
	// Emit runs of same-styled runes so one escape sequence covers many cells.
	for i := 0; i < len(runes); {
		j, on := i, isHit[i]
		for j < len(runes) && isHit[j] == on {
			j++
		}
		st := base
		if on {
			st = hit
		}
		b.WriteString(st.Render(string(runes[i:j])))
		i = j
	}
	return b.String()
}

File: src/finder/list.rs
use std::collections::HashMap;

use ratatui::style::Style;
use ratatui::text::{Line, Span};

use super::command::split_input;
use super::score::{find, match_score, substring_match};
use super::worktree::Mode;
use super::{Cmd, Model, Msg};
use crate::repo::{Branch, PullRequest, Visit};

/// Item is one row: a repository in the main list, a worktree, branch or pull
/// request in the others. label is what is drawn and matched, path is what
/// Enter yields.
#[derive(Debug, Clone, Default)]
pub struct Item {
    pub label: String,
    pub path: String,
    pub score: f64,  // frecency; zero outside the repository list
    pub seen: Visit, // the visit log; empty outside the repository list
    /// branch is the branch a row of the branch list stands for. Such a row
    /// has a path only when the branch is already checked out somewhere.
    pub branch: Branch,
    /// pr is the pull request a row of the pull request list stands for, with
    /// a path the same way.
    pub pr: PullRequest,
}

impl Model {
    pub(super) fn filter(&mut self) {
        self.matched.clear();
        self.note.clear();
        // A command is not a query: the list stays as the query before it left it.
        let value = self.input.value();
        let q = split_input(value.trim())
            .map_or(value.trim().to_string(), |(q, _)| q)
            .trim()
            .to_string();
        // The selection follows the item, not the row number, whenever the
        // query is not making a new ranking statement — a command being typed,
        // or the query being cleared. Clearing it has to hold the selection
        // still, or there is no way to find a repository and then act on it.
        let hold = !self.view_stale && (q == self.query || q.is_empty());
        let held = if hold {
            self.view.get(self.cursor).copied()
        } else {
            None
        };
        self.view_stale = false;

        let view: Vec<usize> = if q.is_empty() {
            (0..self.all.len()).collect()
        } else {
            let lower = q.to_lowercase();
            let labels: Vec<&str> = self.all.iter().map(|it| it.label.as_str()).collect();
            let mut scored: Vec<(usize, f64)> = Vec::new();
            for mt in find(&q, &labels) {
                let rel = labels[mt.index];
                // fuzzy finds candidates; where it landed the characters is its
                // own guess, and a literal hit beats that guess every time.
                let pos = substring_match(rel, &lower).unwrap_or(mt.matched);
                scored.push((mt.index, match_score(rel, &pos)));
                self.matched.insert(mt.index, char_indexes(rel, &pos));
            }
            // Ascending, so the strongest match lands at the bottom; frecency
            // breaks ties between equally good matches.
            scored.sort_by(|a, b| {
                a.1.total_cmp(&b.1)
                    .then(self.all[a.0].score.total_cmp(&self.all[b.0].score))
            });
            scored.into_iter().map(|(i, _)| i).collect()
        };
        self.query = q;
        self.view = self.keep_repo_filters(view);

        if let Some(i) = held.and_then(|h| self.view.iter().position(|&x| x == h)) {
            self.cursor = i;
            return;
        }
        self.cursor = self.view.len().saturating_sub(1);
    }

    /// keep_repo_filters applies status filters to the repository list only.
    /// While the shared scan is in flight, every row stays visible.
    fn keep_repo_filters(&self, view: Vec<usize>) -> Vec<usize> {
        if self.mode != Mode::Repos || !(self.dirty_only || self.unpushed_only) {
            return view;
        }
        let Some(states) = &self.repo_states else {
            return view;
        };
        view.into_iter()
            .filter(|&i| {
                states.get(&self.all[i].path).is_some_and(|state| {
                    (!self.dirty_only || state.dirty > 0)
                        && (!self.unpushed_only || state.unpushed > 0 || state.ahead > 0)
                })
            })
            .collect()
    }

    /// scan_repo_states asks git about every repository at once, off the UI
    /// thread. The result serves both the dirty and unpushed filters.
    pub(super) fn scan_repo_states(&self, busy_tag: u64) -> Cmd {
        let paths: Vec<String> = self.all.iter().map(|it| it.path.clone()).collect();
        let scan = self.state_scan_of.clone();
        Cmd::Task(Box::new(move || {
            Some(Msg::RepoStates(busy_tag, scan(&paths)))
        }))
    }

    /// render_row draws one row, highlighting the characters the query
    /// matched and, when selected, the whole line.
    pub(super) fn render_row(&self, i: usize, selected: bool, width: usize) -> Line<'static> {
        let idx = self.view[i];
        let it = &self.all[idx];
        let (base, hit, marker) = if selected {
            (
                self.st.row_sel,
                self.st.hit_sel,
                Span::styled("▸ ", self.st.marker),
            )
        } else {
            (self.st.row, self.st.hit, Span::raw("  "))
        };
        let empty = Vec::new();
        let mut spans = vec![marker];
        spans.extend(highlight(
            &it.label,
            self.matched.get(&idx).unwrap_or(&empty),
            width.saturating_sub(2),
            base,
            hit,
        ));
        let used: usize = spans.iter().map(Span::width).sum();
        if width > used {
            spans.push(Span::styled(" ".repeat(width - used), base));
        }
        Line::from(spans)
    }
}

/// char_indexes converts byte offsets into char positions, which is what
/// matching and rendering use to track highlights.
fn char_indexes(s: &str, byte_idx: &[usize]) -> Vec<usize> {
    s.char_indices()
        .enumerate()
        .filter(|(_, (b, _))| byte_idx.contains(b))
        .map(|(i, _)| i)
        .collect()
}

/// highlight renders s truncated to width display cells, with the chars at
/// hits in their own style. The tail is kept: the repository name matters more
/// than the host.
fn highlight(s: &str, hits: &[usize], width: usize, base: Style, hit: Style) -> Vec<Span<'static>> {
    if width <= 1 {
        return Vec::new();
    }
    let chars: Vec<char> = s.chars().collect();
    let mut spans = Vec::new();
    let mut off = 0;
    let text_width: usize = chars.iter().map(|c| Span::raw(c.to_string()).width()).sum();
    if text_width > width {
        let suffix_width = width - 1;
        let mut kept_width = 0;
        off = chars.len();
        while off > 0 {
            let char_width = Span::raw(chars[off - 1].to_string()).width();
            if kept_width + char_width > suffix_width {
                break;
            }
            kept_width += char_width;
            off -= 1;
        }
        spans.push(Span::styled("…", base));
    }
    let is_hit: HashMap<usize, bool> = hits
        .iter()
        .filter(|&&h| h >= off)
        .map(|&h| (h - off, true))
        .collect();
    // Runs of same-styled chars, so one span covers many cells.
    let mut i = 0;
    while i < chars.len() - off {
        let on = is_hit.contains_key(&i);
        let mut j = i;
        while j < chars.len() - off && is_hit.contains_key(&j) == on {
            j += 1;
        }
        spans.push(Span::styled(
            chars[off + i..off + j].iter().collect::<String>(),
            if on { hit } else { base },
        ));
        i = j;
    }
    spans
}
